src/main/java/com/ctc/wstx/sr/StreamScanner.java

```diff
@@ -1188,55 +1188,121 @@ protected int resolveSimpleEntity(boolean checkStd)
+                        if (ptr >= inputLen) {
+                            break;
+                        }
+                        c = buf[ptr++];
                     }
                 }
-            } else { // numeric (decimal)
-                while (c != ';') {
-                    if (c <= '9' && c >= '0') {
-                        value = (value * 10) + (c - '0');
-                        // Overflow?
-                        if (value > MAX_UNICODE_CHAR) {
-                            reportUnicodeOverflow();
+                
+                if (pairValue >= 0) {
+                    pairValue = tmpValue;
+                } else {
+                    value = tmpValue;
+                }
+                
+                isValueHighSurrogate = value >= 0xD800 && value <= 0xDBFF;
+                
+                /*
+                 * If resolving entity surrogate pairs enabled and if current entity
+                 * is in range of high surrogate value, try to find surrogate pair 
+                 */
+                if (isValueHighSurrogate && mConfig.isResolvingEntitySurrogatePairs() 
+                        && c == ';' && pairValue < 0 && ptr + 1 < inputLen) {
+                    c = buf[ptr++];
+                    
+                    if (c == '&' && ptr + 1 < inputLen) {
+                        c = buf[ptr++];
+                        
+                        if (c == '#' && ptr + 1 < inputLen) {
+                            c = buf[ptr++];
+                            pairValue = 0;
+                            continue;
+                        } else {
+                            reportNoSurrogatePair(value);
                         }
                     } else {
-                        mInputPtr = ptr; // so error points to correct char
-                        throwUnexpectedChar(c, "; expected a decimal number.");
+                        reportNoSurrogatePair(value);
                     }
-                    if (ptr >= inputLen) {
-                        break;
-                    }
-                    c = buf[ptr++];
+                } else if (isValueHighSurrogate 
+                        && mConfig.isResolvingEntitySurrogatePairs() 
+                        && ptr + 1 >= inputLen) {
+                    reportNoSurrogatePair(value);
                 }
             }
+            
             /* We get here either if we got it all, OR if we ran out of
              * input in current buffer.
              */
             if (c == ';') { // got the full thing
                 mInputPtr = ptr;
-                validateChar(value);
+                
+                if (mConfig.isResolvingEntitySurrogatePairs() && pairValue > 0) {
+                    /*
+                     * If pair value is not in range of low surrogate values, then throw an error
+                     */
+                    if (pairValue < 0xDC00 || pairValue > 0xDFFF) {
+                        reportNoSurrogatePair(value);
```

Comment:
Would be better if we reported both leading surrogate and second non-surrogate entity value?